numba/help/inspector.py

```diff
@@ -0,0 +1,388 @@
+"""
+This file contains `__main__` so that it can be run as a commandline tool.
+
+This file contains function to inspect Numba's support for a given Python
+module or a Python package.
+"""
+from __future__ import print_function
+
+import argparse
+import pkgutil
+import types as pytypes
+
+from numba._version import get_versions
+from numba.targets.registry import cpu_target
+from numba.tests.support import captured_stdout
+
+
+commit = get_versions()['full'].split('.')[0]
+github_url = 'https://github.com/numba/numba/blob/{commit}/{path}#L{firstline}-L{lastline}'
+
+
+def inspect_function(function, target=None):
+    """Return information about the support of a function.
+
+    Returns
+    -------
+    info : dict
+        Defined keys:
+        - "numba_type": str or None
+            The numba type object of the function if supported.
+        - "explained": str
+            A textual descrption of the support.
+        - "source_infos": dict
+            A dictionary containing the source location of each definition.
+    """
+    target = target or cpu_target
+    tyct = target.typing_context
+    # Make sure we have loaded all extensions
+    tyct.refresh()
+    target.target_context.refresh()
+
+    info = {}
+    # Try getting the function type
+    source_infos = {}
+    try:
+        nbty = tyct.resolve_value_type(function)
```

Comment:
Not quite sure where to put this comment but I think it's probably to do with this line... It seems like resolving by module doesn't work particularly well due to aliasing. For example...
```
numpy.ma.core.array()
```
is the NumPy masked array impl which is IIRC not supported at all, and yet this code produces:
```
numpy.ma.core.array()
        defined by <class 'numba.typing.templates.Registry.register_global.<locals>.decorate.<locals>.Template'>
```
and I think this is because it considers functions per module opposed to being API aware. Numba, I think, has all it's functions defined with respect to the package root, so an implementation for `array` can only ever be `np.array`? I also think that this is by convention opposed to enforcement, because one could `from numpy import array` and then write something that matches `array` but its not clear where the `array` came from unless the function is checked against the root. FWIW https://github.com/numba/numba/pull/3942 also wrestles with the same problem .

Reply:
There's also the issue of modules importing functionality from elsewhere and there being repeat definitions as a result, e.g. `np.linalg.linalg.all == np.all`. I'm wondering if this tool should walk the top level module and push discovered functions into a `set`, then subsequent submodules can check to see if a function on their module is already defined and explained elsewhere. This approach is imperfect but may reduce repetition.

Also, perhaps `.tests.*` submodules can be skipped?

Reply:
Changes:

 - `tests` modules are skipped
- fixed a bug that causes top-level package API to not show.
- fixed a bug that causes `numpy.ma.core.array` to be marked as supported (the wrong name was used).
- aliases symbols are now indicated with "alias to: <first occurence>".